synapse/cores/common.py

```diff
@@ -220,6 +236,98 @@ def __init__(self, link, **conf):
 
         s_ingest.IngestApi.__init__(self, self)
 
+    def getModlVers(self, name):
+        '''
+        Retrieve the model version for the given model name.
+
+        Args:
+            name (str): The name of the model
+
+        Returns:
+            (int):  The model version ( linear version number ) or -1
+        '''
+        prop = '.:modl:vers:' + name
+        rows = self.getRowsByProp(prop)
+        if not rows:
+            return -1
+        return rows[0][2]
+
+    def setModlVers(self, name, vers):
+        '''
+        Set the version number for a specific model.
+
+        Args:
+            name (str): The name of the model
+            vers (int): The new (linear) version number
+
+        Returns:
+            (None)
+
+        '''
+        prop = '.:modl:vers:' + name
+
+        with self.getCoreXact() as xact:
+
+            rows = self.getRowsByProp(prop)
+
+            if rows:
+                iden = rows[0][0]
+            else:
+                iden = guid()
+
+            self.setRowsByIdProp(iden,prop,vers)
+            return vers
+
+    def revModlVers(self, name, revs):
+        '''
+        Update a model using a list of (vers,func) tuples.
+
+        Args:
+            name (str): The name of the model
+            refs ([(int,function)]):  List of (vers,func) revision tuples.
+
+        Returns:
+            (None)
+
+        Example:
+
+            def v0(core):
+                addModelStuff(core)
+
+            def v1(core):
+                addMoreStuff(core)
+
+            revs = [ (0,v0), (1,v1) ]
+
+            core.revModlVers('foo', revs)
+
+        Each specified function is expected to update the cortex including data migration.
+        '''
+        if not revs:
+            return
+
+        curv = self.getModlVers(name)
+
+        maxver = revs[-1][0]
+        if maxver == curv:
+            return
+
+        if not self.getConfOpt('rev:model'):
+            raise NoRevAllow(name=name, mesg='add rev:model=1 to cortex url to allow model updates')
+
+        for vers,func in sorted(revs):
+
+            if vers <= curv:
+                continue
+
+            # allow the revision function to optionally return the
+            # revision he jumped to ( to allow initial override )
+            retn = func()
```

Comment:
I think most use cases will be via decorators on a CoreModule, but i'd like to keep the API public to enable odd cases.  :)